### PR TITLE
Apply HFS+ compression on archive extraction

### PR DESCRIPTION
**Pull request by raimue**
This is a continuation from #45.

Ticket: https://trac.macports.org/ticket/36560

### CODE REVIEW

**Comment by RJVB** [comment]
Remind me again why you don't simply rely on `bsdtar --hfsCompression`, for the extraction step? I know that used to extract empty files on ZFS but if that would ever be an issue it has now been fixed in ZFS.

File: src/registry2.0/portimage.tcl [hunk 1]
```diff
@@ -351,45 +351,45 @@ proc extract_archive_to_tmpdir {location} {
...
-                            ui_debug $errmsg
-                            throw MACPORTS "No '$gzip' was found on this system!"
+                            set gzip "bzip2"
                         }
+                    } elseif {[regexp {lz$} ${unarchive.type}]} {
+                        set gzip "lzma"
+                    } elseif {[regexp {xz$} ${unarchive.type}]} {
+                        set gzip "xz"
+                    } else {
+                        set gzip "gzip"
```
Comment (neverpanic) [inline src/registry2.0/portimage.tcl]: While we're changing this anyway, should we run `s/gzip/compression/g`? `set gzip "xz"` just looks weird to me.

**Comment by raimue** [comment]
@RJVB I trusted you when you [said](https://github.com/macports/macports-base/pull/45#discussion_r149611365) that `bsdtar --hfsCompression` raises an error on filesystems that do not support compression.

File: src/registry2.0/portimage.tcl [hunk 2]
```diff
@@ -350,15 +350,28 @@ proc extract_archive_to_tmpdir {location} {
...
-                    ui_debug $errmsg
-                    throw MACPORTS "No '$tar' was found on this system!"
+                # Opportunistic HFS compression. bsdtar will automatically
+                # ignore if on non-HFS filesystem, or if extraction device
+                # is not the same as prefix / applications_dir
+                array set fsinfo [statfs ${extractdir}]
+                if {![catch {macports::binaryInPath bsdtar}] &&
+                    ![catch {exec bsdtar -x --hfsCompression < /dev/null >& /dev/null}] &&
+                    $fsinfo(mntfromname) == [statfs ${macports::prefix} mntfromname] &&
+                    $fsinfo(mntfromname) == [statfs ${macports::applications_dir} mntfromname] &&
```
[the remark this answers is not in the document]
Comment (eborisch) [reply src/registry2.0/portimage.tcl]: Premature optimization? ;)

It is purely to save the additional time required to extract with compression if the compression is going to be lost due to a cross-device move. (This was requested in a comment on the original PR.)

Certainly skipping this check (and supporting code) makes this PR much more consice; all of the statfs code can be pulled out.

**Comment by raimue** [comment]
Okay, reading the old discussion again, apparently only ZFS was faulty. I would keep the check for fstypename only.

**Comment by RJVB** [comment]
>Okay, reading the old discussion again, apparently only ZFS was faulty. I would keep the check for fstypename only.

I'd say that libarchive is at fault but indeed, only on ZFS (to my knowledge and re: filesystems we're likely to encounter on Mac).

The fix they applied in ZFS is to decompress on the fly, and then possibly apply whatever transparent compression is set on the dataset...

As to cross-device moving: there are ways to preserve the compression in that scenario. Rsync can do so optionally, and the host's ditto command will too.

R.

**Comment by raimue** [comment]
libarchive seems [to check](https://github.com/libarchive/libarchive/blob/339356e373bfd47370954a4882386cd24d2f3e13/libarchive/archive_write_disk_posix.c#L696-L713) the underlying filesystem by trying to apply `fchflags(..., UF_COMPRESSED)`. As `/usr/bin/bsdtar` does not support `--hfsCompression` anyway, it is probably safe to assume this will always work correctly and we can drop the filesystem type checks.

I do not worry too much about preserving HFS+ compression when moving files from `${prefix}/var/macports/software/mpextract...` into `${prefix}`. It is unlikely someone configures their system this way as it already comes with a major performance drop on activation. If need be, that can still be looked into later.

**Comment by RJVB** [comment]
> I do not worry too much about preserving HFS+ compression when moving files from `${prefix}/var/macports/software/mpextract...` into `${prefix}`. It is unlikely someone configures their system this way

Why? Anyone with $prefix on a different volume than /Applications will be affected, and /Applications/MacPorts can get quite large too (880Mb for me, 371Mb compressed with level zip-8).

> as it already comes with a major performance drop on activation. If need be, that can still be looked into later.

Do you think so? I don't find it to be an issue, certainly not compared to creating the xz-compressed archives. Of course I'm too poor (or is that cheap?) to invest in a large enough SSD and the volume holding my $prefix has an interesting amount of free-space fragmentation by now, so activating a larger port always takes a noticeable amount of time.

In a sense this might be something to worry a lot about. People who opt to use hfs compression on top of having their $prefix on a dedicated volume might actually run that configuration because they have a small boot disk. Not compressing /Applications/MacPorts would be a bit of a missed opportunity there, even if the absolute gain might not be as impressive as what can be attained on $prefix.

R.

**Comment by eborisch** [comment]
Perhaps that should be a separate discussion: how to maintain compression when activating from a different device.

For the bulk of users (laptop especially), I'd wager it's one device for everything, or at least everything MP related. This is a big win for those users (again, especially laptop), in terms of space used by MP. Putting in lots of checks for a small percentage of users (where it's not _breaking_ things by not checking, but perhaps not time/CPU optimal) makes the code more fragile in general, and the time and complexity might be better spent and reduced, respectively, by looking at the activation step separately.

My original version https://github.com/macports/macports-base/pull/45/commits/d715cb34dbeed6c91c341c373bbbd86f67502263 just checked for if the flag was supported by bsdtar and then said "let's do this." I was fine with it then, and am still fine with that now. ;)

**Comment by raimue** [comment]
Dropped all of statfs. HFS+ compression will be applied with bsdtar if it supports the --hfsCompression flag.

Basically back to the original submission of @eborisch, but the tar check has been fixed. I rephrased the commit message, but kept you as author.

Preserving HFS+ compression across device renames in `_activate_file` is out of scope of this pull request and can be addressed afterwards. Most users will only run from a single filesystem anyway, so this is not a blocker.

**Comment by janstary** [comment]
How much slower is it to untar when performing the HFS compression? About six times slower.
```
$ du -hs file.tar
749M    file.tar

$ time /opt/local/bin/bsdtar xf file.tar
$ time /opt/local/bin/bsdtar xf file.tar
[...]
real    0m6.762s
user    0m0.061s
sys     0m1.232s

$ time /opt/local/bin/bsdtar --hfsCompression -xf file.tar
$ time /opt/local/bin/bsdtar --hfsCompression -xf file.tar
[...]
real    0m48.513s
user    0m27.129s
sys     0m3.573s
```

I don't care how much space is saves, drop it.

**Comment by cjones051073** [comment]
On 18/04/18 10:08, Jan Starý wrote:
> How much slower is it to untar when performing the HFS compression? 
> About six times slower.
> 
> |$ du -hs file.tar 749M file.tar $ time /opt/local/bin/bsdtar xf file.tar 
> $ time /opt/local/bin/bsdtar xf file.tar [...] real 0m6.762s user 
> 0m0.061s sys 0m1.232s $ time /opt/local/bin/bsdtar --hfsCompression -xf 
> file.tar $ time /opt/local/bin/bsdtar --hfsCompression -xf file.tar 
> [...] real 0m48.513s user 0m27.129s sys 0m3.573s |
> 
> I don't care how much space is saves, drop it.

That is just your opinion. You are welcome to it but others might 
disagree, and prefer the disk space savings over the CPU time.

I suggest if it isn't there already an option should be added to disable 
(or enable, depending on the default) this option in the macports 
configuration. Then users can decide for themselves which they prefer.

**Comment by raimue** [comment]
I benchmarked this on a MacBook Pro (Retina, Mid 2012), with macOS 10.12 Sierra, with the root filesystem being on a "APPLE SSD SM256E" formatted as "Mac OS Extended (Journaled, Encrypted)". The largest port archive I have to test is texlive-latex-extra, which also contains many small files, so it should be a good test candidate. libarchive for bsdtar with --hfsCompression is installed as well.

```
$ port -q installed libarchive texlive-latex-extra
  libarchive @3.3.2_1 (active)
  texlive-latex-extra @44430_0+doc (active)
$ ls -lh $(port -q location texlive-latex-extra)
-rw-r--r--  1 root  admin   418M Jun 19  2017 /opt/local/var/macports/software/texlive-latex-extra/texlive-latex-extra-44430_0+doc.darwin_16.noarch.tbz2
```

Unfortunately, I don't know how to reliable sum up the size on disk to measure the difference in file size, as `du` thinks these compressed files take 0 bytes. Therefore I only took the time for activation.

Without hfsCompression, master b200459e83cab1924130ef78628bde16b0e5503e:
```
$ sudo port -Nf deactivate texlive-latex-extra
...
$ time sudo port -N activate texlive-latex-extra
...
real	1m3.553s
user	0m51.842s
sys	0m15.657s
$ port -q contents texlive-latex-extra | xargs ls -lO | awk '{print $5}' | sort | uniq -c
10185 -
```

With hfsCompression, master b200459e83cab1924130ef78628bde16b0e5503e + the patches from this pull request applied on top:
```
$ sudo port -Nf deactivate texlive-latex-extra
...
$ time sudo port -N activate texlive-latex-extra
...
real	1m5.655s
user	1m11.009s
sys	0m9.592s
$ port -q contents texlive-latex-extra | xargs ls -lO | awk '{print $5}' | sort | uniq -c
  63 -
10122 compressed
```

I see no impact by activating HFS+ Compression with this patch on my system.

**Comment by eborisch** [comment]
@raimue, I'm not sure that's the right testing technique. I see a savings of 85MB (16%) for texlive-latex-extra, see below.

Tiny files (<k) aren't really the best candidates. In fact, they're almost the worst case because they take a long time to process (lots of files == lots of overhead) and you don't get much out of it.

Large text files (header files) are. Compilers, or ports with large help / HTML files.

To test:
 * make sure libarchive is installed
 * check disk usage `df -k /opt/local/`
 * deactivate & re-activate a large port (llvm / gcc / clang)
 * check disk usage again. If everything is working, you should have more space available. 

Here's a test script:
```sh
#!/bin/bash
function space { df -m /opt/local | awk '/2/{print $4}'; }
function echo3 { echo "$*" >&3; }

exec 3>&1
exec &> size_test.log

echo3 "Deactivating $1 (and libarchive)"
sudo port -f deactivate libarchive $1

echo3 "Activating $1 (without compression)"
sudo port activate $1
SIZE_A=$(space)

echo3 "Deactivating $1 and activating libarchive"
sudo port -f deactivate $1
SIZE_UNC=$(($(space) - SIZE_A))
sudo port activate libarchive
SIZE_B=$(space)

echo3 "Activating $1 (with compression)"
sudo port activate $1
SIZE_COMP=$((SIZE_B - $(space)))

exec 1>&3

echo "For port $1:"
echo "Compressed size: ${SIZE_COMP}MB; Uncompressed size: ${SIZE_UNC}MB"
echo "Savings: $((SIZE_UNC - SIZE_COMP))MB"
```

And tested with llvm-6.0 , clang-6.0, and texlive-latex-extra:

```
$ ./test_size.sh llvm-6.0
Deactivating llvm-6.0 (and libarchive)
Activating llvm-6.0 (without compression)
Deactivating llvm-6.0 and activating libarchive
Activating llvm-6.0 (with compression)
For port llvm-6.0:
Compressed size: 49MB; Uncompressed size: 159MB
Savings: 110MB

./test_size.sh clang-6.0
Deactivating clang-6.0 (and libarchive)
Activating clang-6.0 (without compression)
Deactivating clang-6.0 and activating libarchive
Activating clang-6.0 (with compression)
For port clang-6.0:
Compressed size: 122MB; Uncompressed size: 370MB
Savings: 248MB

$ ./test_size.sh texlive-latex-extra
Deactivating texlive-latex-extra (and libarchive)
Activating texlive-latex-extra (without compression)
Deactivating texlive-latex-extra and activating libarchive
Deactivating texlive-latex-extra (with compression)
For port texlive-latex-extra:
Compressed size: 467MB; Uncompressed size: 552MB
Savings: 85MB
```

And yes, it is slower, @janstary, but saving 360MB on clang+llvm is worth it for some people.

**Comment by raimue** [comment]
> Tiny files (<k) aren't really the best candidates. In fact, they're almost the worst case because they take a long time to process (lots of files == lots of overhead) and you don't get much out of it.

I meant many small files are the best candidate to show the overhead it will have on extraction. In my test, activation did not take significantly longer. For measuring time, it is important to know the details of the Mac and the version of macOS this test is run on, whether the disk is a SSD, using HFS+ or APFS as all of that could make a difference for this.

Testing with `df -m` is not really accurate, as other processes will write to the disk at the same time (log files, daemon state files, etc.). But it should be good enough to show a trend.

**Comment by RJVB** [comment]
On Wednesday April 18 2018 08:22:52 Eric A. Borisch wrote:

> Absolutely, but it gets away from trying to look at apparent vs. on-disk vs. metadata overhead discussions... Test when idle for best results; re-test for some level of assurance.

How often do you activate a port when the system is idle? Measuring a 6x slow-down because of HFS compression may seem impressive under such circumstances, but that number could be smaller in real life (because uncompressed activation gets much slower). 

I've been using a HFS compression patch for years, on a mid 2011 2.6Ghz i7 with a traditional HDD. The extra activation time has never been an issue because I the ports that take most time to activate are also those where I have the least reason to do this regularly.

R.

**Comment by eborisch** [comment]
> How often do you activate a port when the system is idle?

Ok @RJVB, Test when the system is _otherwise_ idle. ;)

And the timing is going to really vary by CPU and disk resources. (number of cores? lbzip2 installed? spinning rust? SSD? etc, etc.). Fast CPU and slow disk will favor this compression patch. Slow CPU and fast disk will be the reverse. But everyone [1] should get the benefit of reduced size.

On my (iMac, 4-core 8-thread, SSD; so fast CPU and disk) system and this patch, activating texlive-latex-extra with both lbzip and libarchive active is faster than activating with them inactive. (The speedup from lbzip2 more than counteracts the slowdown from hfs compression.)

[1] Except those where the extraction partition != active partition, as already discussed above.

**Comment by pmetzger** [comment]
I presume that the goal would be to have this available if someone turned it on? (Rather than on by default?)

**Comment by RJVB** [comment]
My 2 cents say make it on by default for new installs (new users won't have a reference against which to notice a slow-down).

**Comment by eborisch** [comment]
I would suggest on by default, and an option to force off.

For most users of Macports, the activation time (the only real drawback) is a small portion of most installs, when you consider port processing time + fetch time, and especially (if needed) build will dwarf activation time, even with compression. Making the footprint smaller on disk (with minimal runtime penalty; I’ve been unable to measure any) is an almost universal good, with the potential exception of storage vendors.

Users frequently de-/re-activating ports for testing purposes would be one use case for disabling. Also users with ZFS or other filesystems / split extract & active mountpoints.

**Comment by janstary** [comment]
I have to admit my case was special: the `file.tar` is a tarball of my `~/Downloads/`,
the majorty of which, I realize now, was a huge zip file. So the untarring was mostly
spending time trying to HFS compress a zip file.

(I don't know how HFS compression works inside, but apparently it can give
terrible results on the _opposite_ of "many small files".)

On an actual port image, it is much better:

```
$ du -hs `port -q location boost`
51M	/opt/local/var/macports/software/boost/boost-1.66.0_0+no_single+no_static+python27+regex_match_extra.darwin_17.x86_64.tbz2

$ cd /tmp

$ time /opt/local/bin/bsdtar -mxf `port -q location boost`
real    0m46.583s
user    0m24.738s
sys     0m13.157s
$ du -hs ./opt/
447M    ./opt/
$ rm -rf ./opt

$ time /opt/local/bin/bsdtar --hfsCompression -mxf `port -q location boost`
real    0m1.321s
user    0m0.824s
sys     0m0.307s
$ du -hs ./opt
56M    ./opt/
$ rm -rf ./opt
``` 

I repeated each about ten times, on an otherwise idle system;
the values above are the closest to the average.

So, without the compression, it untars a 51M image into a 56M hierarchy in about 01:32,
and with the compression, it untars the same image into a 447M hierarchy in about 00:47.
That's an impressive compression, for the price of about twice as much time.

Trying on "many small files subject to good compression":

```
$ cd /tmp
$ /opt/local/bin/bsdtar cf file.tar /opt/local/share/doc/boost/
$ du -hs file.tar
249M	file.tar

$ rm -rf ./opt
$ time /opt/local/bin/bsdtar -mxf file.tar
real	0m10.326s
user	0m1.227s
sys	0m7.924s
$ du -hs ./opt
279M	./opt/

$ rm -rf ./opt
$ time /opt/local/bin/bsdtar --hfsCompression -mxf file.tar
real	0m17.212s
user	0m6.636s
sys	0m6.466s
$ du -hs ./opt
 42M	./opt/
```

That's a very nice compression, in about 180% of the time.

**Comment by eborisch** [comment]
> du does not take that into account at all and is therefore not reliable for compressed files.

Which, for all its inadequacies in terms of impact of other IO on the system, I was aiming to "get around" with df.

**Comment by RJVB** [comment]
> Depending on the compressed size, the data can either be stored in a resource fork, or in an extended attribute called "com.apple.decmpfs". `du` does not take that into account at all and is therefore not reliable for compressed files.

Actually, it isn't wrong for those small files (which compress to 4096 bytes max). That content gets stored in the file attribute, and thus not *in* the file.

If "gdu" doesn't already take those into account it shouldn't be hard to add that and defend the patch upstream. Same for the "gls" command (ls -sl should list the actual and the on-disk sizes though it seems the 10.9 /usr/bin/ls fails on decmpfs'ed files too).

Afsctool has an option to report on compressed files and directories; otherwise the Finder's GetInfo command will report the actual and on-disk sizes too.

R.

**Comment by janstary** [comment]
> Standard HFS compression is simply zip (not sure what level by default). Zipping an already zipped file is indeed expensive; libarchive could check whether the file being extracted is compressible before wasting cycles on something that might take an inordinate amount of time and/or lead to something that's actually bigger than the original.

The data being uncompressed from the `tar.bz2`  port image
and being compressed into the HFS zip at the same time
will be binaries, libraries, manpages and the like.
Apparently, they can be compressed to quite an extent.
If we checked (how?) whether the content "is compressible",
it would no longer be transparent. The compression either happens or not,
but we don't want to care (right?), we just pass the `--hfscompression` option.

**Comment by janstary** [comment]
> Do not get fooled by du. As I said above, for some files du will report 0 bytes. Depending on the compressed size, the data can either be stored in a resource fork, or in an extended attribute called "com.apple.decmpfs". du does not take that into account at all and is therefore not reliable for compressed files.

@raimue If something as elementary as `du` does not work on these files,
that alone is (in my book anyway) reason to consider it a broken feature and not use it.

Question:  how much space does this package take on disk?
Answer: we don't know, and have no way of knowing.

@eborisch Using `df` to find the size occuied by a given package is a crude workaround at best.
The obvious `port -q content libsamplerate | xargs du -hsc` comes to mind,
but using HFS compression makes that unusable.

**Comment by raimue** [comment]
@janstary You have to complain to Apple about that. Almost all files on a standard macOS installation are compressed. Fixing `du` is up to Apple and out of scope for MacPorts.

**Comment by eborisch** [comment]
>Question: how much space does this package take on disk?
>Answer: we don't know, and have no way of knowing.

> @eborisch Using df to find the size occuied by a given package is a crude workaround at best.

Yes, it is crude @janstary, but you're confusing the issue here, and "[we] have no way of knowing" is a bit of a stretch. Perhaps "we have no know of conveniently knowing exactly". Feel free to create a fork of `du` to fix deficiencies, but please comment on the topic of the PR: reducing the on-disk footprint of a MacPorts installation.

At the scale (hundreds of MB for some packages) we're trying to improve things, `df` is sufficient to show that it works. In fact, the used space reported by `df` lines up with the used space reported by the finder for the volume, which is probably what most users are interested in. (How much space do i have left for photos?)

>Apparently, they can be compressed to quite an extent. If we checked (how?) whether the content "is compressible", it would no longer be transparent. The compression either happens or not, but we don't want to care (right?), we just pass the --hfscompression option.

It appears that libarchive does this (don't write the compressed version if it is larger than the original) block-by-block: https://github.com/libarchive/libarchive/blob/339356e373bfd47370954a4882386cd24d2f3e13/libarchive/archive_write_disk_posix.c#L1244

**Comment by pmetzger** [comment]
> I would suggest on by default, and an option to force off.

For many users, this will mean a significant performance hit for something they're not that interested in. Perhaps we can start at least with the option as off by default?

**Comment by pmetzger** [comment]
BTW, am I correct in noting that any new or APFS using Mac will not notice this because APFS has no such transparent compression options?

**Comment by eborisch** [comment]
>For many users, this will mean a significant performance hit for something they're not that interested in. 

Citation needed. Tests so far here have shown in general a minimal performance hit during extraction.  Almost all files on your system compressed by Apple, BTW, so the run-time hit was clearly deemed tolerable. (Indeed, for compressible files, it can speed things up as less disk access is required.)

 >BTW, am I correct in noting that any new or APFS using Mac will not notice this because APFS has no such transparent compression options?

Widely held belief, but false. I'm running APFS and see the significant reductions in size noted above.

You can also check just about any Apple application, right-click to 'Show package contents', find a compressible file of decent size (most .plist files) and get info. You should see Size: XXX,YYY bytes (NN KB on disk, compressed)

**Comment by eborisch** [comment]
>Does your APFS systm perform HFS compression?

`bsdtar --hfsCompression` creates compressed files on APFS filesystems, yes.

**Comment by eborisch** [comment]
>Note that this is only transparent in decompression, but requires explicit setup when writing files. Therefore feature comparisons for filesystems might not give it a check mark for "transparent compression".

A well done early dive [1] into the FS noted it as missing, but 'likely to show up'. I think this got picked up elsewhere in the HFS+ vs. APFS comparisons and often lost the 'likely to show up' part.

Also of note from that article, relevant to the discussion today: "[...]I even recalled my job interview with Be in 2000 when they talked about how compression actually improved overall performance since data I/O is far more expensive than computation (obvious now, but novel then)."

[1] http://dtrace.org/blogs/ahl/2016/06/19/apfs-part1/

**Comment by RJVB** [comment]
> Fixing `du` is up to Apple and out of scope for MacPorts.

"Their" du, yes. The du in port:coreutils not so much. And if "fixing" (incl. updating) stuff provided by Apple was really out of scope for MacPorts, couldn't we stop providing a whole bunch of ports that "only" provide newer versions of libraries and stuff Apple consider good enough? ;)

> >BTW, am I correct in noting that any new or APFS using Mac will not notice
> >this because APFS has no such transparent compression options?> 
> Widely held belief, but false.

Indeed: APFS support the same compression options as HFS. (Which might mean that decompression is not implemented in the filesystem driver but just above it in the kernel.)

> Does your APFS systm perform HFS compression?

No, the filesystem doesn't perform compression. Otherwise we wouldn't be having this discussion...

> they talked about how compression actually improved overall performance

I can confirm that from practical experience. I have an SD card I bought for my 2011 MBP because it supposedly would give me 64Gb or so of "built-in" SSD. Turns out my MBP only supports "slow" SDs; the only way to make it bearable is to run ZFS on it with transparent lz4 compression.

R

**Comment by eborisch** [comment]
It seems we’re getting wrapped around the axle here. Fixing `du` is out of the scope of this pull request. (Its broken-ness does not change to answer to the question “is this PR is useful/functional?”)

Salient points, this PR:
 * Works on HFS+ and APFS. (Where in the stack the compression “happens” may be an interesting academic discussion, but again, not critical or of interest to most users.)
 * bsdtar automatically disables/ignores the change on unsupported FS.
 * Makes installed software use less of a precious resource (drive space). Sometimes significantly so — see my clang and llvm results above.
 * Works block-by-block on files (tested) to compress portions that can be compressed.
 * Can be slower to extract, especially if the installation contains primarily “incompressible” (.jpeg, .tar.*, etc) files.
 * Can also be faster to extract, if CPUs are outpacing storage. (Frequent with HDD, ? on SSD.)
 * Decompression (reading an installed file) is faster than the compression step; most of a fresh install of OSX & applications is compressed, as a reference point for any performance-once-active concerns.
 * Does not work (but doesn’t break anything) in the case of extraction and destination directiories on different devices.  * A configuration knob to disable is available for those with drive space to burn. ;)

I’d suggest merging it with a default “on” unless someone has a strong reason not to based on live testing.

**Comment by janstary** [comment]
> Where in the stack the compression “happens” may be 
an interesting academic discussion, but again, not critical or of interest to   
most users.

That's not an academic discussion at all. If a *userland application* needs to be concerned
about *how blocks are written to disk*, then something is fundamentaly wrong.

**Comment by eborisch** [comment]
>That's not an academic discussion at all. If a userland application needs to be concerned about how blocks are written to disk, then something is fundamentaly wrong.

Open a radr? “Implement transparent encryption” or “Use ZFS?” https://developer.apple.com/bug-reporting/

**Comment by raimue** [comment]
I think we discussed this long enough, we are starting to digress. It is out of scope to fix Apple's filesystems or systems tools. We have to work with what macOS provides.

HFS+ compression is now enabled by default if bsdtar with support for --hfsCompression is available. It can be disabled using `hfscompression no` in macports.conf. Please give us feedback when you disable it, so we might be able to determine system parameters when it should not be enabled by default.

**Pull request merged**